### ISSUE TITLE
Uniform handling of y_train in forecasting performance metrics

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/sktime#712]
**Is your feature request related to a problem? Please describe.**
Some forecasting performance metrics require `y_train`, e.g. [MASE](https://robjhyndman.com/papers/mase.pdf). This adds some complication to higher-level functionality that expects a common interface for metrics, like the `evaluate` function or `ForecastingGridSearchCV`, and in unit testing (see #672).

**Current problem**
This currently fails because `y_train` is not passed internally when calling `scoring`. 
```python
from sktime.forecasting.all import *
from sktime.forecasting.model_evaluation import evaluate

y = load_airline()
f = NaiveForecaster()
cv = SlidingWindowSplitter()
scoring = MASE()
out = evaluate(f, cv, y, scoring=scoring)
```

**Possible solutions**
1. Change interface for all performance metrics to optionally accept `y_train`, but only those that require it use it. This requires wrapping metrics from scikit-learn.  
2. Add case distinctions in higher-level functionality to separately handle those metrics that require `y_train` and those that do not. This requires adding a `requires_y_train` attribute to metric classes.
3. Adapt metrics interface at run time to, making case distinctions inside adapter, exposing uniform interface to higher-order functionality (suggested by @fkiraly). This also requires adding a `requires_y_train` attribute to metric classes.

**Describe the solution you'd like**
```python
from sktime.forecasting.all import *
from sktime.forecasting.model_evaluation import evaluate

y = load_airline()
fh = np.arange(1, 10)
y_train, y_test = temporal_train_test_split(y, fh=fh)
f = NaiveForecaster()
f.fit(y_train)
y_pred = f.predict(fh)

# uniform interface
scoring = MASE()
scoring.requires_y_train = True
scoring = check_scoring(scoring)
scoring(y_test, y_pred, y_train)
>>> 3.577770878609128

scoring = sMAPE()
scoring.requires_y_train = False
scoring = check_scoring(scoring)
scoring(y_test, y_pred, y_train)
>>> 0.1780237534499896
```

Here's a rough implementation of the adapter-based solution:
```python
class _MetricAdapter:
    """
    Adapter for performance metrics to uniformly handle 
    y_train requirement of some metrics.
    """

    def __init__(self, metric):
        # wrap metric object
        self.metric = metric
        
    def __call__(self, y_true, y_pred, y_train, *args, **kwargs):
        """Compute metric, uniformly handling those metrics that 
        require `y_train` and those that do not.
        """
        
        # if y_train is required, pass it on
        if self.metric.requires_y_train:
            return self.metric(y_true, y_pred, y_train, *args, **kwargs)
        
        # otherwise, ignore y_train
        else:
            return self.metric(y_true, y_pred, *args, **kwargs)     
        
    def __getattr__(self, attr):
        # delegate attribute queries to the wrapped metric object
        return getattr(self.metric, attr)

    def __repr___(self):
        return repr(self.metric)

    
def _adapt_scoring(scoring):
    """Helper function to adapt scoring to uniformly handle y_train requirement"""
    return MetricAdapter(scoring)


def check_scoring(scoring):
    """
    Validate `scoring` object.

    Parameters
    ----------
    scoring : object
        Callable metric object.
    
    Returns
    -------
    scoring : object 
        Validated `scoring` object, or sMAPE() if `scoring` is None.
    
    Raises
    ------
    TypeError
        If `scoring` is not a callable object.
    """
    from sktime.performance_metrics.forecasting import sMAPE
    from sktime.performance_metrics.forecasting._classes import MetricFunctionWrapper
    
    if scoring is None:
        return sMAPE()

    if not callable(scoring):
        raise TypeError("`scoring` must be a callable object")

    valid_base_class = MetricFunctionWrapper
    if not isinstance(scoring, valid_base_class):
        raise TypeError(f"`scoring` must inherit from `{valid_base_class.__name__}`")

    return _adapt_scoring(scoring)
```

**Comment by RNKuhns** [comment]
@mloning I've been reading through the proposal. Wouldn't another option be to just incorporate the conditional logic in the callable aspect of the metrics to start callable to start? Since I'm relatively new to considering the impact of these choices on a module API, I'm Interested to hear you thought process on deciding between incorporating the logic in MetricFunctionWrapper to start instead of adapting it at run time.

I was also wondering about the `__call__` part of the `_adapter`. By defining  `__call__(self, y_true, y_pred, y_train, *args, **kwargs)` is our goal to require every metric call to pass y_train? 

Why not just let `y_train` be passed in the *args? We could assume that when it is necessary it is the first of any optional args that are passed.  Let me know what you think the pros and cons of this approach are versus explicitly requiring y_train being passed.  

Also note that some of the metrics require a benchmark forecast (these might not be easy to work into cross-validation unless we allow users to provide a benchmark or a benchmark method as part of an orchestrater). That being said, we can accommodate those metrics in MetricFunctionWrapper the same way y_train is accommodated pretty easily. 

Something like the code below would incorporate the conditional callability in MetricFunctionWrapper instead of adapting it at run time. I've also added the accommodation of y_pred_benchmark. 

```python
class MetricFunctionWrapper:
    def __init__(
        self, fn, name=None, greater_is_better=False, requires_y_train=False, 
        requires_y_pred_benchmark=False
    ):
        self.fn = fn
        self.name = name if name is not None else fn.__name__
        self.greater_is_better = greater_is_better
        self.requires_y_train=requires_y_train
        self.requires_y_pred_benchmark=requires_y_pred_benchmark

    def __call__(self, y_true, y_pred, *args, **kwargs):
        # if y_train is required, pass it on
        if self.requires_y_train:
            y_train = args[0]
            if len(args) >  1:
                args = args[1:]
            else:
                args=()
            return self.fn(y_true, y_pred, y_train, *args, **kwargs)

        # if y_pred_benchmark is required, pass it on
        elif self.requires_y_pred_benchmark:
            y_pred_benchmark= args[0]
            if len(args) >  1:
                args = args[1:]
            else:
                args=()
            return self.fn(y_true, y_pred, y_pred_benchmark, *args, **kwargs)
        
        # otherwise, ignore y_train and y_pred_benchmark
        else:
            return self.fn(y_true, y_pred, *args, **kwargs) 
```

Another alternative would be to do something like what you had originally but make the y_train parameter handle either the training data or a benchmark predictions. 

```python
class MetricFunctionWrapper:
    def __init__(
        self, fn, name=None, greater_is_better=False, requires_y_train=False, 
        requires_y_pred_benchmark=False
    ):
        self.fn = fn
        self.name = name if name is not None else fn.__name__
        self.greater_is_better = greater_is_better
        self.requires_y_train=requires_y_train
        self.requires_y_pred_benchmark=requires_y_pred_benchmark

    def __call__(self, y_true, y_pred, y_train_or_pred_benchmark, *args, **kwargs):
        # if y_train or prediction benchmark is required, pass it on
        if self.requires_y_train or self.requires_y_pred_benchmark:
            return self.fn(y_true, y_pred, y_train_or_pred_benchmark, *args, **kwargs)
        
        # otherwise, ignore y_train and y_pred_benchmark
        else:
            return self.fn(y_true, y_pred, *args, **kwargs) 
```

**Comment by mloning** [comment]
Hi @RNKuhns, that sounds like a good solution too, may even be simpler. Just to be clear, you mean to add a new hidden method to the scoring base class that we call inside higher-level functionality? Something along the following lines?

```python

class BaseMetric:

    def __call__(...):
        ... # regular user call

    def _score(y_true, y_pred, y_train, ...):
        ... # uniform handling of y_train inside higher-level functionality
```

**Comment by RNKuhns** [comment]
@mloning that sounds a simple solution to me and I like the name `_score` much better than `_apply_scoring` that I suggested.

**Comment by mloning** [comment]
@RNKuhns ok let's go for it :)

**Comment by RNKuhns** [comment]
@mloning and @fkiraly I saw #857 popped up. I should have a draft pull request related to this fix in the next day or two. 

I've got the logic worked out to use `y_train` when necessary based on the `_tags` assigned to the metric. Just wanted to confirm the solution really quick, because I believe another suggestion might have come up when we closed out #801. 

Based on conversation above, one option is: 
```
class BaseMetric:

    def __call__(... # regular arguments needed by the metric):
        ... # regular user call

    def _score(y_true, y_pred, y_train, ...):
        ... # uniform handling of y_train inside higher-level functionality using _tagsabout y_train being required
```
Where higher-level functionality like `evaluate` would always use the `_score(y_true, y_pred, y_train, ...)` method, but any regular user call would just require the user to pass whatever the usual call signature would be for the metric (e.g. we don't require the regular user call to include y_train when it isn't necessary). 

The other option that I remember someonebeing mentioning (by @fkiraly I think) during the wrap-up of #801 was the possibility of a keyword argument for y_train that defaults to None (I interpreted this as being a keyword argument in `__call__`). In that case the solution would be something like the following: 

```
class BaseMetric:

    def __call__(y_true, y_pred, y_train = None):
        ... # logic would use _tags to determine if y_train should be used in metric or not
```

I can see pros and cons of both approaches. First option is nice that it doesn't expose users to keyword parameters that don't make sense for a given metric (e.g. why is y_train keyword parameter an option for mean_squared_error), but provides ability for things like `evaluate` to use `_score()` so we can include things like `MASE` in evaluation and tuning. 

On the other hand, the second approach is nice in that the user interface when a user calls the class and when higher-level functionality like `evaulate` use the metric class are the same.  

The two approaches have a trade-off between having the same `__call__` arguments for every metric versus potentially confusing users by having `y_train` as keyword argument for metrics it is not needed for. Either should be workable for users with proper documentation (which we'll include). What are your thoughts from an API design perspective? 

Does having different `__call__` signatures violate our API design principles, even if we have the uniform

**Comment by fkiraly** [comment]
@RNKuhns, I strongly favour option 2, because option 1 would make call of the templated classes non-uniform, and result in error if passing `y_train` to the call.

The general principle is that it should be easy to loop over objects of the same type, or to exchange one object for the other.
In option 1 this is not possible via the `__call__` interface point, it will result in errors whenever you switch a non-`y_train` with a requires-`y_train` instance, either way round.

**Comment by fkiraly** [comment]
PS: you can achieve option 2 by catching kwargs and simply ignoring them - no need for extra internal handling of `y_train` in that case.

**Comment by fkiraly** [comment]
Regarding option 1: I can see the merits, but my concern is that it´s a bit brittle - we have `_score` in a sense duplicating `__call__`, so there's not one way to do things ("python zen"), and the way is not obvious. This is because you need to know that you need to use `_score` when you need a "uniform interface call" and that's a concept not obvious to the "typical user".

The orange flag is that we would have to explicitly tell @aiwalter that that is what he needs to use in `evaluate`, and whenever people are not told it seems that there is risk for non-compliance with the interface.

**Comment by RNKuhns** [comment]
@fkiraly thanks for the input. That does make a lot of sense. One quick question on catching the `kwargs`. Do we want to allow `__call__` to accept kwargs? I thought we generally tried to avoid them and use explicit parameters instead?

**Comment by fkiraly** [comment]
good question - @mloning, what's your opinion for `__call__`? I think we have kwargs in constructors?

**Comment by mloning** [comment]
Thanks for bringing this up again. I initially favoured option 2, I now agree with @fkiraly and prefer option 1. 

The signature should then be: `__call__(self, y_true, y_pred, y_train=None)` where some metric require `y_train` and raise an error if not given and others don't require it and simply ignore it if given. (Does this imply that all performance metric functions should also accept `y_train`? I somehow still feel less comfortable with that.)

The point about keyword arguments in `__call__` is less about keyword arguments per se and more about configuration arguments for the metrics (e.g. `symmetric=True`) which should be in the constructor, not the `__call__` method.

**Comment by fkiraly** [comment]
>Does this imply that all performance metric functions should also accept y_train? I somehow still feel less comfortable with that.

@mloning, I think the wrapper (class that makes metric class from function) could take care of that in case we don't want functions to have that argument - though I think we can still catch this with kwargs even there.

**Comment by RNKuhns** [comment]
@mloning No problem bringing it back up. I am with you in terms of not being as comfortable with the functions in `_functions.py` all requiring `y_train`. 

Also thanks for the clarification on kwargs, it helps to hear about the background on these things so I can keep that in mind on future work! I definately get that we want to keep hyperparameters/config args in the constructor and the `__call__` method. I had been thinking that ruled out kwargs in general. 

If we are good to use kwargs for things that don't fall into that hyper-parameter/confi args bucket, then I'm with @fkiraly and think we use that to catch `y_train` when it is passed to function. 

Higher-level functionality like `evaluate` would just always pass `y_train`.

I'll start implementing and can make any tweaks once I get a draft PR.

**Comment by RNKuhns** [comment]
@mloning and @fkiraly I've started a draft PR (#858). 

I'll be updating how the performance_metric tests yet, but wanted to start the PR to get some initial feedback.

**Comment by mloning** [comment]
Closed by #858
